### ISSUE TITLE
As a user, I don't want an evaluation to fail with a subscriber error on publishing paths created

### ISSUE THREAD

**Issue by epag** [NOAA-OWP/wres#117]
---

Author Name: **James** (James)
Original Redmine Issue: 117034, https://vlab.noaa.gov/redmine/issues/117034
Original Date: 2023-06-07

---

Given an evaluation of system test scenario703 in #89538-1145
When it completes
Then I expect it to succeed, not fail with a subscriber error and eventual @SubscriberTimedOutException@

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **James** (James)
Original Date: 2023-06-07T14:11:36Z

---

```
    2023-06-07T11:33:00.993+0000 ERROR EvaluationSubscriber While processing a statistics message, encountered an exception that will stop the subscriber.
    java.lang.IllegalStateException: End size 2928 is less than fixed size 2950
    	at java.base/java.util.stream.Nodes$FixedNodeBuilder.end(Nodes.java:1232)
    	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
    	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510)
    	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
    	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
    	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
    	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
    	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
    	at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
    	at wres.events.subscribe.EvaluationConsumer.publishCompletionState(EvaluationConsumer.java:717)
    	at wres.events.subscribe.EvaluationConsumer.markEvaluationSucceeded(EvaluationConsumer.java:341)
    	at wres.events.subscribe.EvaluationSubscriber.registerEvaluationCompleteIfConsumptionComplete(EvaluationSubscriber.java:822)
    	at wres.events.subscribe.EvaluationSubscriber.lambda$getStatisticsListener$2(EvaluationSubscriber.java:681)
    	at org.apache.qpid.jms.JmsMessageConsumer.deliverNextPending(JmsMessageConsumer.java:801)
    	at org.apache.qpid.jms.JmsMessageConsumer$MessageDeliverTask.run(JmsMessageConsumer.java:863)
    	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    	at org.apache.qpid.jms.util.QpidJMSThreadFactory$1.run(QpidJMSThreadFactory.java:102)
    	at java.base/java.lang.Thread.run(Thread.java:833)
    2023-06-07T11:33:00.998+0000 ERROR EvaluationSubscriber Message subscriber nZhDcmPxmA16iQ06S5iJL8exBOg has been flagged as failed without the possibility of recovery.
    java.lang.IllegalStateException: End size 2928 is less than fixed size 2950
    	at java.base/java.util.stream.Nodes$FixedNodeBuilder.end(Nodes.java:1232)
    	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
    	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510)
    	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
    	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
    	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
    	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
    	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
    	at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
    	at wres.events.subscribe.EvaluationConsumer.publishCompletionState(EvaluationConsumer.java:717)
    	at wres.events.subscribe.EvaluationConsumer.markEvaluationSucceeded(EvaluationConsumer.java:341)
    	at wres.events.subscribe.EvaluationSubscriber.registerEvaluationCompleteIfConsumptionComplete(EvaluationSubscriber.java:822)
    	at wres.events.subscribe.EvaluationSubscriber.lambda$getStatisticsListener$2(EvaluationSubscriber.java:681)
    	at org.apache.qpid.jms.JmsMessageConsumer.deliverNextPending(JmsMessageConsumer.java:801)
    	at org.apache.qpid.jms.JmsMessageConsumer$MessageDeliverTask.run(JmsMessageConsumer.java:863)
    	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    	at org.apache.qpid.jms.util.QpidJMSThreadFactory$1.run(QpidJMSThreadFactory.java:102)
    	at java.base/java.lang.Thread.run(Thread.java:833)

```

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **James** (James)
Original Date: 2023-06-07T14:26:00Z

---

Looks like a very low % concurrency bug (we haven't seen it before over a period of years) whereby a list of paths created is being streamed for publication, but the paths are still being updated. Publication should occur after all writing has completed and all paths have been recorded, but apparently in some very low % situation that is not the case. Likely a problem with @EvaluationConsumer::isComplete@ w/r respecting the paths mutated, perhaps this:

```<code class="java">
    private void acceptStatisticsMessageInner( Statistics statistics, String groupId, String messageId )
    {
        // Accept the incremental types
        this.execute( () -> this.addPathsWritten( this.getConsumer()
                                                      .apply( statistics ) ) );

        // Accept the grouped types
        if ( Objects.nonNull( groupId ) )
        {
            OneGroupConsumer<Statistics> groupCon = this.getGroupConsumer( groupId );

            // Get the statistics for grouped consumption
            Statistics groupedStatistics = this.getStatisticsForGroupedConsumption( statistics, groupId, messageId );

            // May trigger group completion and consumption
            this.execute( () -> groupCon.accept( messageId, groupedStatistics ) );

            this.closeGroupIfComplete( groupCon );
        }

        // Record consumption
        this.consumed.incrementAndGet();
</code>
```

In other words, the execution of the writing (@this.execute( () -> this.addPathsWritten( foo )@) is async to the recording of consumption complete/increment, @this.consumed.incrementAndGet();@

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **James** (James)
Original Date: 2023-06-07T14:28:47Z

---

Either way, if it happens once per N years, it's most likely going to be a speculative fix and not super urgent.
